apps/condo/domains/common/components/TrackingContext.tsx

```diff
@@ -0,0 +1,210 @@
+        if (!domainName) {
+            return ''
+        }
+        const domain = domainName.charAt(0).toUpperCase() + domainName.slice(1)
+        let domainSuffix = 'Index'
+        let domainPostfix = ''
+
+        switch (isDetail) {
+            case '[id]':
+                domainSuffix = 'Detail'
+                break
+            case 'create':
+                domainSuffix = 'Create'
+                break
+        }
+
+        if (suffix) {
+            switch (suffix) {
+                case 'update':
+                    domainPostfix = 'Update'
+                    break
+                case 'pdf':
+                    domainPostfix = 'Pdf'
+                    break
+                default:
+                    // convert string from domain-related-page to DomainRelatedPage
+                    domainPostfix = suffix.replace(/(-\w|^\w)/g, (_, g) => g.slice(-1).toUpperCase())
+                    break
+            }
+        }
+
+        return `${domain}${eventType}${domainSuffix}${domainPostfix}`
+    }
+
+    return {
+        eventProperties,
+        userProperties,
+        logEvent,
+        logEventTo,
+        instrument,
+        getEventName,
+    }
+}
+
+const TrackingProvider: React.FC = ({ children }) => {
+    const { user } = useAuth()
+    const { link } = useOrganization()
+    const router = useRouter()
+
+    const trackingProviderValueRef = useRef<ITrackingContext>({
+        trackerInstances: TRACKING_INITIAL_VALUE.trackerInstances,
+        userProperties: {},
+        eventProperties: {
+            page: {
+                path: router.asPath,
+            },
+        },
+    })
+
+    const routeChangeStart = (url) => {
+        trackingProviderValueRef.current.eventProperties.page.path = url
+    }
+
+    useEffect(() => {
+        // Init all instances of trackers only on client side rendering
+        if (typeof window !== 'undefined') {
+            // Page path changed -> change value at context object
+            router.events.on('routeChangeStart', routeChangeStart)
+            Object.values(trackingProviderValueRef.current.trackerInstances).forEach(trackerInstance => trackerInstance.init())
+        }
+
+        return () => {
+            router.events.off('routeChangeStart', routeChangeStart)
+        }
+    }, [])
+
+    // Collect user & organization related data to slice custom groups based on given attributes
+    useEffect(() => {
+        if (user) {
+            trackingProviderValueRef.current.userProperties = pick(user, TRACKING_USER_FIELDS)
+
+            if (link) {
+                trackingProviderValueRef.current.userProperties['role'] = get(link, 'role.name')
+                trackingProviderValueRef.current.userProperties['organization'] = get(link, 'organization.name')
+            }
+        }
+    }, [user, link])
+
+    return (
+        <TrackingContext.Provider value={{ ...trackingProviderValueRef.current }}>
```

Comment:
creating object here will cause rerenderings, better to use `React.useMemo` outside of `return` for cloning this

Reply:
Remove object spread